### PR TITLE
Land PostgresClient that is backed by a ConnectionPool as SPI

### PR DESCRIPTION
**Pull request by fabianfett**

### CODE REVIEW

File: Sources/PostgresNIO/Pool/PostgresClient.swift [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+            public var maximumConnectionHardLimit: Int = 10
+
+            public var maxConsecutivePicksFromEventLoopQueue: UInt8 = 16
+
+            public var connectionIdleTimeout: Duration = .seconds(60)
+
+            public var keepAliveFrequency: Duration = .seconds(30)
+
+            public var keepAliveQuery: PostgresQuery = "SELECT 1;"
```
Comment (gwynne) [inline Sources/PostgresNIO/Pool/PostgresClient.swift:30]: Can we get some doc comments for these?

File: Sources/PostgresNIO/Pool/PostgresClient.swift [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+            public static func require(_ sslContext: NIOSSL.TLSConfiguration) -> Self {
+                self.init(.require(sslContext))
+            }
+        }
+
+        public struct Server: Sendable {
+            /// The server to connect to
+            ///
+            /// - Default: localhost
+            public var host: String = "localhost"
```
Comment (gwynne) [inline Sources/PostgresNIO/Pool/PostgresClient.swift:87]: No UDS support?

File: Sources/PostgresNIO/Pool/PostgresClient.swift [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        ConnectionRequest.ID,
+        PostgresKeepAliveBehavor,
+        PostgresClientMetrics,
+        ContinuousClock
+    >
+
+    let pool: Pool
+
+    public init(configuration: Configuration, eventLoopGroup: EventLoopGroup, backgroundLogger: Logger) throws {
+        let connectionConfig = try PostgresConnection.Configuration(configuration)
```
Comment (gwynne) [inline Sources/PostgresNIO/Pool/PostgresClient.swift:129]: Variant that uses the singleton ELG?

File: Tests/IntegrationTests/PostgresClientTests.swift [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+extension PostgresClient.Configuration {
+    static func makeTestConfiguration() -> PostgresClient.Configuration {
+        var clientConfig = PostgresClient.Configuration()
+        clientConfig.pool.minimumConnectionCount = 0
+        clientConfig.pool.maximumConnectionSoftLimit = 8*4
+        clientConfig.pool.maximumConnectionHardLimit = 12*4
+        clientConfig.pool.keepAliveFrequency = .seconds(5)
+        clientConfig.pool.connectionIdleTimeout = .seconds(15)
+
+        clientConfig.server.host = env("POSTGRES_HOSTNAME") ?? "localhost"
```
Comment (gwynne) [inline Tests/IntegrationTests/PostgresClientTests.swift:47]: Need to respect `POSTGRES_PORT` too.

File: Tests/IntegrationTests/PostgresClientTests.swift [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import Atomics
+
+@available(macOS 13.0, iOS 16.0, tvOS 16.0, watchOS 9.0, *)
+final class PostgresClientTests: XCTestCase {
+
+    func testGetConnection() async throws {
+        var mlogger = Logger(label: "test")
+        mlogger.logLevel = .debug
+        let logger = mlogger
+        let eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 8)
```
Comment (gwynne) [inline Tests/IntegrationTests/PostgresClientTests.swift:14]: Would `System.coreCount` be more representative here?

File: Sources/PostgresNIO/Pool/PostgresClient.swift [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+@available(macOS 13.0, iOS 16.0, tvOS 16.0, watchOS 9.0, *)
+struct PostgresKeepAliveBehavor: ConnectionKeepAliveBehavior {
+    var keepAliveFrequency: Duration?
+    var query: PostgresQuery
+    var logger: Logger
+
+    init(keepAliveFrequency: Duration?, logger: Logger) {
+        self.keepAliveFrequency = keepAliveFrequency
+        self.query = "SELECT 1;"
```
Comment (gwynne) [inline Sources/PostgresNIO/Pool/PostgresClient.swift:186]: Shouldn't this be configurable via the corresponding client config field?

**Review by gwynne — CHANGES_REQUESTED** [review]

File: Tests/IntegrationTests/PostgresClientTests.swift (hunk 5 above)
Comment (fabianfett) [reply Tests/IntegrationTests/PostgresClientTests.swift:15]: This is mainly to ensure that concurrent access is really, really safe.

File: Sources/PostgresNIO/Pool/ConnectionFactory.swift [hunk 7]
```diff
@@ -141,6 +141,7 @@ final class ConnectionFactory: Sendable {
                 break
 
             case .produce:
+                // TBD: we might want to consider moving this of the concurrent executor
```
Comment (gwynne) [inline Sources/PostgresNIO/Pool/ConnectionFactory.swift:144]:
```suggestion
                // TBD: we might want to consider moving this off the concurrent executor
```

File: Sources/PostgresNIO/Pool/PostgresClient.swift [hunk 8]
```diff
@@ -1,8 +1,36 @@
...
 import NIOSSL
+import Atomics
 import Logging
 import _ConnectionPoolModule
 
+/// A Postgres client that is backed by an underlying connection pool. Use ``Configuration`` to change the client's
+/// behavior.
+///
+/// > Important:
+/// The client can only lease connections, if the user is running the client's ``run()`` method in a long running task:
```
Comment (gwynne) [inline Sources/PostgresNIO/Pool/PostgresClient.swift:11]:
```suggestion
/// The client can only lease connections if the user is running the client's ``run()`` method in a long running task:
```

**Comment by fabianfett** [comment]
Crashing compiler toolchain in nightly. Ignore for now.

File: Sources/PostgresNIO/Pool/ConnectionFactory.swift [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+        case .require(let tlsConfiguration):
+            let sslContext = try await self.getSSLContext(for: tlsConfiguration)
+            tls = .require(sslContext)
+        case .disable:
+            tls = .disable
+        }
+
+        var connectionConfig: PostgresConnection.Configuration
+        switch config.endpointInfo {
+        case .bindUnixDomainSocket(let path):
```
Comment (gwynne) [inline Sources/PostgresNIO/Pool/ConnectionFactory.swift:70]: Come to think of it, how about the "existing `Channel`" thing too?

File: Sources/PostgresNIO/Pool/PostgresClient.swift [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+            observabilityDelegate: .init(logger: backgroundLogger),
+            clock: ContinuousClock()
+        ) { (connectionID, pool) in
+            let connection = try await factory.makeConnection(connectionID, pool: pool)
+
+            return ConnectionAndMetadata(connection: connection, maximalStreamsOnConnection: 1)
+        }
+    }
+
+    public func withConnection<Result>(_ closure: (PostgresConnection) async throws -> Result) async throws -> Result {
```
Comment (gwynne) [inline Sources/PostgresNIO/Pool/PostgresClient.swift:279]: Doc comments.

**Review by gwynne — CHANGES_REQUESTED** [review]

File: Sources/PostgresNIO/Pool/ConnectionFactory.swift (hunk 9 above)
Comment (fabianfett) [reply Sources/PostgresNIO/Pool/ConnectionFactory.swift:70]: How do we want to use a single Channel for multiple connections. We will need to come up with a custom factory. That is a valuable goal. But not today.

**Review by gwynne — APPROVED** [review]

**Pull request merged**